File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
#Bottom-up 방식
def solution(triangle):
    dp = [row[:] for row in triangle]
    n = len(triangle)
    
    for i in range(n-2,-1,-1):
        for j in range(len(triangle[i])):
            dp[i][j] += max(dp[i+1][j],dp[i+1][j+1])

    return dp[0][0]

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;

class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        int n = triangle[triangle.length-1].length;
        int[][] dp = new int[n][n];
        
        dp[0][0] = triangle[0][0];
        
        for(int i = 1; i<n; i++) {
            dp[i][0] = dp[i-1][0] + triangle[i][0];
        }
        
        for(int i = 1; i<n; i++) {
            for(int j = 1; j<triangle[i].length; j++) {
                dp[i][j] = Math.max(dp[i-1][j], dp[i-1][j-1]) + triangle[i][j];
            }
        }
        
        for(int i = 0; i<n; i++) {
            answer = Math.max(dp[n-1][i], answer);
        }
        
        return answer;
    }
}